R/filechoose.R

```diff
@@ -2,376 +2,395 @@
-  
+  if (missing(pattern)) {
+    pattern <- ""
+  } else if (is.function(pattern)) {
+    pattern <- pattern()
+  }
+
   function(dir, root) {
-    currentRoots <- if(class(roots) == 'function') roots() else roots
-    
-    if (is.null(names(currentRoots))) stop('Roots must be a named vector or a function returning one')
+    currentRoots <- if (class(roots) == "function") roots() else roots
+
+    if (is.null(names(currentRoots))) stop("Roots must be a named vector or a function returning one")
     if (is.null(root)) root <- names(currentRoots)[1]
-    
-    ## drop paths with only "" to avoid // 
+
+    ## drop paths with only "" to avoid //
     dropEmpty <- function(x) x[!vapply(x, function(x) nchar(x) == 0, FUN.VALUE = logical(1))]
     fulldir <- file.path(currentRoots[root], dropEmpty(dir))
-    
+
     dropEmpty <- function(x) x[!vapply(x, function(x) nchar(x) == 0, FUN.VALUE = logical(1))]
-    fulldir <- do.call('file.path', as.list(dropEmpty(c(currentRoots[root], dir))))
+    fulldir <- do.call("file.path", as.list(dropEmpty(c(currentRoots[root], dir))))
     writable <- as.logical(file.access(fulldir, 2) == 0)
-    files <- list.files(fulldir, all.files=hidden, full.names=TRUE, no..=TRUE)
+    files <- list.files(fulldir, all.files = hidden, full.names = TRUE, no.. = TRUE)
     if (!is.null(restrictions) && length(files) != 0) {
       if (length(files) == 1) {
-        keep <- !any(sapply(restrictions, function(x) {grepl(x, files, fixed=T)}))
+        keep <- !any(sapply(restrictions, function(x) {
+          grepl(x, files, fixed = T)
+        }))
       } else {
-        keep <- !apply(sapply(restrictions, function(x) {grepl(x, files, fixed=T)}), 1, any)
+        keep <- !apply(sapply(restrictions, function(x) {
+          grepl(x, files, fixed = T)
+        }), 1, any)
       }
       files <- files[keep]
     }
     fileInfo <- file.info(files)
     fileInfo$filename <- basename(files)
     fileInfo$extension <- tolower(file_ext(files))
-    fileInfo$mtime <- format(fileInfo$mtime, format='%Y-%m-%d-%H-%M')
-    fileInfo$ctime <- format(fileInfo$ctime, format='%Y-%m-%d-%H-%M')
-    fileInfo$atime <- format(fileInfo$atime, format='%Y-%m-%d-%H-%M')
+    fileInfo$mtime <- format(fileInfo$mtime, format = "%Y-%m-%d-%H-%M")
+    fileInfo$ctime <- format(fileInfo$ctime, format = "%Y-%m-%d-%H-%M")
+    fileInfo$atime <- format(fileInfo$atime, format = "%Y-%m-%d-%H-%M")
     if (!is.null(filetypes)) {
-      matchedFiles <- tolower(fileInfo$extension) %in% tolower(filetypes) & fileInfo$extension != ''
+      matchedFiles <- tolower(fileInfo$extension) %in% tolower(filetypes) & fileInfo$extension != ""
       fileInfo$isdir[matchedFiles] <- FALSE
-      fileInfo <- fileInfo[matchedFiles | fileInfo$isdir,]
+      fileInfo <- fileInfo[matchedFiles | fileInfo$isdir, ]
+    }
+    if (nchar(pattern) > 0) {
+      matchedFiles <- try(grepl(pattern, fileInfo$filename), silent = TRUE)
+      if (!is(matchedFiles, "try-error")) {
```

Comment:
methods is generally a horrible thing to depend on (without reason) - here you can simply use `!inherits(matchedFiles, 'try-error')`